File: src/test/scala/com/mobikok/ssp/data/streaming/OM.java
//package com.mobikok.ssp.data.streaming;
//
//import com.fasterxml.jackson.core.type.TypeReference;
//import com.fasterxml.jackson.databind.DeserializationFeature;
//import com.fasterxml.jackson.databind.ObjectMapper;
//import com.fasterxml.jackson.databind.PropertyNamingStrategy;
//import com.fasterxml.jackson.databind.SerializationFeature;
//import com.fasterxml.jackson.databind.cfg.MapperConfig;
//import com.fasterxml.jackson.databind.introspect.AnnotatedMethod;
//import org.apache.commons.io.IOUtils;
//import org.apache.spark.sql.Dataset;
//import org.apache.spark.sql.Row;
//import org.slf4j.Logger;
//import org.slf4j.LoggerFactory;
//
//import java.io.File;
//import java.io.FileInputStream;
//import java.io.FileOutputStream;
//import java.io.IOException;
//import java.sql.ResultSet;
//import java.sql.ResultSetMetaData;
//import java.sql.SQLException;
//import java.util.ArrayList;
//import java.util.HashMap;
//import java.util.List;
//import java.util.Map;
//
///**
// * Created by Administrator on 2017/6/20.
// */
//public class OM {
//
//	private static Logger LOG = LoggerFactory.getLogger(OM.class);
//
//	private static ObjectMapper INDENT_OUTPUT_OBJECT_MAPPER = new ObjectMapper().configure(
//			SerializationFeature.INDENT_OUTPUT, true).disable(
//			DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
//
//	private static ObjectMapper LOWER_CASE_OBJECT_MAPPER = new ObjectMapper().disable(
//			DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
//			.setPropertyNamingStrategy(new PropertyNamingStrategy() {
//				public String nameForSetterMethod(MapperConfig<?> config,
//												  AnnotatedMethod method, String defaultName) {
//					return defaultName.toLowerCase();
//				}
//			});
//
//	private static ObjectMapper OBJECT_MAPPER = new ObjectMapper().disable(
//			DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES);
//
//	public static String toJOSN(Object o) {
//		try {
//			return INDENT_OUTPUT_OBJECT_MAPPER.writeValueAsString(o);
//		} catch (Exception e) {
//			throw new RuntimeException(e);
//		}
//	}
//
//	public static <T> T toBean(String json, Class<T> clazz) {
//		try {
//			return INDENT_OUTPUT_OBJECT_MAPPER.readValue(json, clazz);
//		} catch (Exception e) {
//			throw new RuntimeException(e);
//		}
//	}
//
//	public static <T> T toBean(String json, TypeReference<T> clazz) {
//		try {
//			return INDENT_OUTPUT_OBJECT_MAPPER.readValue(json, clazz);
//		} catch (Exception e) {
//			throw new RuntimeException(e);
//		}
//	}
//
//	public static <T> T convert(Object o, TypeReference<T> clazz) {
//		try {
//			return INDENT_OUTPUT_OBJECT_MAPPER.convertValue(o, clazz);
//		} catch (Exception e) {
//			throw new RuntimeException(e);
//		}
//	}
//
//	public static <T> T convert(Object o, Class<T> clazz) {
//		try {
//			return INDENT_OUTPUT_OBJECT_MAPPER.convertValue(o, clazz);
//		} catch (Exception e) {
//			throw new RuntimeException(e);
//		}
//	}
//
//	public static <T> T readBeanFromJSONFile(String filePath, TypeReference<T> typeReference){
//
//		try {
//
//			File f = new File(filePath);
//			if(!f.exists()) {
//				throw new RuntimeException("JSON File '"+filePath+"' not exists !!");
//			}
//
//			List<String> lines = IOUtils.readLines(new FileInputStream(filePath));
//
//			StringBuffer buff = new StringBuffer();
//			for(String s : lines) {
//				buff.append("\n").append(s);
//			}
//
//			LOG.warn("readBeanFromJSONFile: \n" +buff.toString());
//
//			String json = buff.toString().replace("\n", "").replaceAll("	", "    ");
//			return com.mobikok.ssp.data.streaming.util.OM.toBean(json, typeReference);
//
//		} catch (Exception e) {
//			throw new RuntimeException("Read JSON file '"+filePath+"' fail: ",e);
//		}
//
//	}
//
//	public static void writeBeanToJSONFile(String filePath, Object bean){
//
//		try {
//			String json = com.mobikok.ssp.data.streaming.util.OM.toJOSN(bean);
//			File f = new File(filePath);
//			if(!f.exists()) {
//				f.createNewFile();
//			}
//			IOUtils.write(json, new FileOutputStream(f));
//
//		} catch (IOException e) {
//			throw new RuntimeException(e);
//		}
//
//	}
//
//	public static <T> List<T> lowerCaseFieldNameAssembleAs(ResultSet rs, Class<T> calzz) {
//
//		List<T> result = new ArrayList<T>();
//		try {
//			while (rs.next()) {
//				Map<String, Object> o = new HashMap<String, Object>();
//
//				ResultSetMetaData md = rs.getMetaData();
//				int c = md.getColumnCount();
//
//				for (int i = 1; i <= c; i++) {
//					String n = rs.getMetaData().getColumnLabel(i);
//					o.put(n, rs.getObject(n));
//				}
//				result.add(LOWER_CASE_OBJECT_MAPPER.convertValue(o, calzz));
//			}
//		} catch (SQLException e) {
//			throw new RuntimeException("遍历JDBC ResultSet异常：", e);
//		}
//
//		return result;
//	}
//
//	public static <T> List<T> assembleAs(ResultSet rs, Class<T> calzz) {
//
//		List<T> result = new ArrayList<T>();
//		try {
//			while (rs.next()) {
//				Map<String, Object> o = new HashMap<String, Object>();
//
//				ResultSetMetaData md = rs.getMetaData();
//				int c = md.getColumnCount();
//
//				for (int i = 1; i <= c; i++) {
//					String n = rs.getMetaData().getColumnLabel(i);
//					o.put(n, rs.getObject(n));
//				}
//				result.add(OBJECT_MAPPER.convertValue(o, calzz));
//			}
//		} catch (SQLException e) {
//			throw new RuntimeException("遍历JDBC ResultSet异常：", e);
//		}
//
//		return result;
//	}
//
////	public static Dataset<Row> assembleAsDataFrame(ResultSet rs) {
////
////		try {
////			ResultSetMetaData md = rs.getMetaData();
////			int cs = md.getColumnCount();
////
////			for(int i = 0; i < cs; i++) {
////				String n = md.getColumnName(i);
////				int t = md.getColumnType(i);
////				if(Types.INTEGER == t) {
////
////				}
////
////
////				StructField("timestamp", LongType) :: Nil
////			}
////
////			list = assembleAs(rs, List.class);
////		} catch (SQLException e) {
////			e.printStackTrace();
////		}
////
////	}
//
//	public static <T> List<T> lowerCaseFieldNameAssembleAs(ResultSet rs, TypeReference<T> calzz) {
//
//		List<T> result = new ArrayList<T>();
//		try {
//			while (rs.next()) {
//				Map<String, Object> o = new HashMap<String, Object>();
//
//				ResultSetMetaData md = rs.getMetaData();
//				int c = md.getColumnCount();
//
//				for (int i = 1; i <= c; i++) {
//					String n = rs.getMetaData().getColumnLabel(i);
//					o.put(n, rs.getObject(n));
//				}
//				result.add((T)LOWER_CASE_OBJECT_MAPPER.convertValue(o, calzz));
//			}
//		} catch (SQLException e) {
//			throw new RuntimeException("遍历JDBC ResultSet异常：", e);
//		}
//
//		return result;
//	}
//
//}

File: src/main/scala/com/mobikok/ssp/data/streaming/util/OM.scala
package com.mobikok.ssp.data.streaming.util

import java.io.{File, FileInputStream, FileOutputStream, IOException}
import java.sql.{ResultSet, ResultSetMetaData, SQLException, Types}
import java.text.SimpleDateFormat
import java.util

import com.fasterxml.jackson.core.`type`.TypeReference
import com.fasterxml.jackson.databind.cfg.MapperConfig
import com.fasterxml.jackson.databind.introspect.AnnotatedMethod
import com.fasterxml.jackson.databind.{DeserializationFeature, ObjectMapper, PropertyNamingStrategy, SerializationFeature}
import com.mobikok.ssp.data.streaming.util.MySqlJDBCClientV2.Callback
import org.apache.commons.io.IOUtils
import org.apache.spark.sql.hive.HiveContext
import org.apache.spark.sql.types.DecimalType.DoubleDecimal
import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.{DataFrame, Dataset, Row, SQLContext, SparkSession}
import org.apache.spark.sql.types._

/**
  * Created by Administrator on 2017/6/20.
  */
object OM {

  val INDENT_OUTPUT_OBJECT_MAPPER = new ObjectMapper().setDateFormat(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")).configure(SerializationFeature.INDENT_OUTPUT, true).disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)
  val OBJECT_MAPPER = new ObjectMapper().setDateFormat(new SimpleDateFormat("yyyy-MM-dd HH:mm:ss")).configure(SerializationFeature.INDENT_OUTPUT, false).disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES)

  def toJOSN(o: Any) : String ={
    toJOSN(o, true)
  }
  def toJOSN(o: Any, indentOutput: Boolean) : String ={
    if(indentOutput) {
      INDENT_OUTPUT_OBJECT_MAPPER.writeValueAsString(o)
    }else {
      OBJECT_MAPPER.writeValueAsString(o)
    }
  }

  def toBean[T] (json: String, clazz: Class[T]) : T ={
    INDENT_OUTPUT_OBJECT_MAPPER.readValue(json, clazz)
  }

  def toBean[T] (json: String, clazz: TypeReference[T]) : T ={
    INDENT_OUTPUT_OBJECT_MAPPER.readValue(json, clazz)
  }

  def convert[T] (o: Object, clazz: TypeReference[T]) : T ={
    INDENT_OUTPUT_OBJECT_MAPPER.convertValue(o, clazz)
  }

  def convert[T] (o: Object, clazz: Class[T]) : T ={
    INDENT_OUTPUT_OBJECT_MAPPER.convertValue(o, clazz)
  }


  def parseStructType(rs: ResultSet): StructType ={
    var structType = new StructType
    val md = rs.getMetaData
    val cs = md.getColumnCount
    val fs = Seq[StructField]()
    var i = 1
    while (i <= cs) {
      val n = md.getColumnLabel(i)
      val t = md.getColumnType(i)

      if (Types.TINYINT == t) {
        structType = structType.add(n, IntegerType)
      }
      else if (Types.SMALLINT == t) {
        structType = structType.add(n, IntegerType)
      }
      else if (Types.INTEGER == t) {
        structType = structType.add(n, IntegerType)
      }
      else if (Types.BIGINT == t) {
        structType = structType.add(n, LongType)
      }
      else if (Types.FLOAT == t || Types.REAL == t) {
        structType = structType.add(n, FloatType)
      }
      else if (Types.DOUBLE == t) {
        structType = structType.add(n, DoubleType)
      }
      else if (Types.DATE == t) {
        structType = structType.add(n, DateType)
      }
      else if (Types.TIME == t) {
        structType = structType.add(n, DateType)
      }
      else if (Types.TIMESTAMP == t) {
        structType = structType.add(n, TimestampType)
      }
      else if (Types.VARCHAR == t) {
        structType = structType.add(n, StringType)
      }
      else if(Types.LONGVARCHAR == t) {
        structType = structType.add(n, StringType)
      }
      else if(Types.LONGNVARCHAR == t) {
        structType = structType.add(n, StringType)
      }
      else if(Types.NCHAR  == t) {
        structType = structType.add(n, StringType)
      }
      else if(Types.NVARCHAR == t) {
        structType = structType.add(n, StringType)
      }
      else if (Types.CHAR == t) {
        structType = structType.add(n, StringType)
      }
      else if(Types.DECIMAL == t) {
        structType = structType.add(n, DecimalType(30, 15))
      }
      else {
        throw new RuntimeException("Undefined can handle column type: "
          + md.getColumnTypeName(i) +" ("+t+")" + " name: " + md.getColumnName(i) + " label: " + md.getColumnLabel(i))
      }
      i += 1
    }
    structType
  }


  private val LOG: org.apache.log4j.Logger = org.apache.log4j.Logger.getLogger(OM.getClass)

  private val LOWER_CASE_OBJECT_MAPPER: ObjectMapper = new ObjectMapper().disable(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES).setPropertyNamingStrategy(new PropertyNamingStrategy() {
    override def nameForSetterMethod (config: MapperConfig[_], method: AnnotatedMethod, defaultName: String): String = defaultName.toLowerCase
  })

  def readBeanFromJSONFile[T] (filePath: String, typeReference: TypeReference[T]): T = try {
    val f: File = new File(filePath)
    if (!f.exists) throw new RuntimeException("JSON File '" + filePath + "' not exists !!")
    val lines: util.List[_] = IOUtils.readLines(new FileInputStream(filePath))
    val buff: StringBuffer = new StringBuffer
    import scala.collection.JavaConversions._
    for (s <- lines) {
      buff.append("\n").append(s)
    }
    LOG.warn("readBeanFromJSONFile: \n" + buff.toString)
    val json: String = buff.toString.replace("\n", "").replaceAll("	", "    ")
    OM.toBean(json, typeReference)
  } catch {
    case e: Exception =>
      throw new RuntimeException("Read JSON file '" + filePath + "' fail: ", e)
  }

  def writeBeanToJSONFile (filePath: String, bean: Any): Unit = {
    try {
      val json: String = OM.toJOSN(bean)
      val f: File = new File(filePath)
      if (!f.exists) f.createNewFile
      IOUtils.write(json, new FileOutputStream(f))
    } catch {
      case e: IOException =>
        throw new RuntimeException(e)
    }
  }

  def assembleAsFromLowerCaseFieldName[T] (rs: ResultSet, beanClass: Class[T]): util.List[T] = {
    val result: util.List[T] = new util.ArrayList[T]
    try
        while (rs.next) {
          val o: util.Map[String, AnyRef] = new util.HashMap[String, AnyRef]
          val md: ResultSetMetaData = rs.getMetaData
          val c: Int = md.getColumnCount
          var i: Int = 1
          while (i <= c) {
            val n: String = rs.getMetaData.getColumnLabel(i)
            o.put(n, rs.getObject(n))

            i += 1;
          }
          result.add(LOWER_CASE_OBJECT_MAPPER.convertValue(o, beanClass))
        }
    catch {
      case e: SQLException =>
        throw new RuntimeException("遍历JDBC ResultSet异常：", e)
    }
    result
  }

  import scala.collection.JavaConversions._

  def assembleAsDataFrame (rs: ResultSet, sqlContext: SQLContext): DataFrame = {
    import sqlContext.implicits._

    val rows: java.util.List[Row] = assembleAsRow(rs)

    val md = rs.getMetaData
    val c = md.getColumnCount
    val fs = new Array[String](c)
    var i = 1
    while (i <= c) {
      fs(i - 1) = md.getColumnLabel(i)
      i += 1
    }
    val sc = parseStructType(rs)
    sqlContext.createDataFrame(rows, sc)
  }

// TEST
//  case class Person(name:String,age:Int)
//  def rddToDFCase(sparkSession : SparkSession): Unit = {
//    //导入隐饰操作，否则RDD无法调用toDF方法
//    import sparkSession.implicits._
//    val peopleRDD = sparkSession.sparkContext
//      .textFile("file:/E:/scala_workspace/z_spark_study/people.txt",2)
//      .map( x => x.split(",")).map( x => /*Person*/(x(0),x(1).trim().toInt))
//
//    peopleRDD.toDF()
//  }


  def assembleAsRow (rs: ResultSet): java.util.List[Row] = {
    val result = new java.util.ArrayList[Row]()
    try
        while (rs.next) {
//          val o: util.Map[String, AnyRef] = new util.HashMap[String, AnyRef]
          val md: ResultSetMetaData = rs.getMetaData
          val c: Int = md.getColumnCount
          var i: Int = 1
          var row = Row()
          val fs = new Array[Any](c)
          while (i <= c) {
            val n: String = rs.getMetaData.getColumnLabel(i)
            //o.put(n, rs.getObject(n))
//            row :+= rs.getObject(n)
            fs(i - 1) = rs.getObject(i)

            i += 1;
          }
          result.add(Row(fs:_*))
//          result.add(Row.fromSeq(seq))
//          result.add(OBJECT_MAPPER.convertValue(o, beanClass))
        }

    catch {
      case e: SQLException =>
        throw new RuntimeException("遍历JDBC ResultSet异常：", e)
    }
    result
  }

  def assembleAs[T] (rs: ResultSet, beanClass: Class[T]): util.List[T] = {
    val result: util.List[T] = new util.ArrayList[T]
    try
        while (rs.next) {
          val o: util.Map[String, AnyRef] = new util.HashMap[String, AnyRef]
          val md: ResultSetMetaData = rs.getMetaData
          val c: Int = md.getColumnCount
          var i: Int = 1
          while (i <= c) {
            val n: String = rs.getMetaData.getColumnLabel(i)
            o.put(n, rs.getObject(n))

            i += 1;
          }
          result.add(OBJECT_MAPPER.convertValue(o, beanClass))
        }

    catch {
      case e: SQLException =>
        throw new RuntimeException("遍历JDBC ResultSet异常：", e)
    }
    result
  }

  def assembleAsFromLowerCaseFieldName[T] (rs: ResultSet, beanCalss: TypeReference[T]): util.List[T] = {
    val result: util.List[T] = new util.ArrayList[T]
    try
        while (rs.next) {
          val o: util.Map[String, AnyRef] = new util.HashMap[String, AnyRef]
          val md: ResultSetMetaData = rs.getMetaData
          val c: Int = md.getColumnCount
          var i: Int = 1
          while (i <= c) {
            val n: String = rs.getMetaData.getColumnLabel(i)
            o.put(n, rs.getObject(n))

            i += 1;
          }
          result.add(LOWER_CASE_OBJECT_MAPPER.convertValue(o, beanCalss).asInstanceOf[T])
        }
    catch {
      case e: SQLException =>
        throw new RuntimeException("遍历JDBC ResultSet异常：", e)
    }
    result
  }
}
